cmd/chartsvc/handler.go

```diff
@@ -61,34 +64,97 @@ type rel struct {
 	Links selfLink    `json:"links"`
 }
 
-// listCharts returns a list of charts
-func listCharts(w http.ResponseWriter, req *http.Request) {
+type meta struct {
+	TotalPages int `json:"totalPages"`
+}
+
+// getPageNumberAndSize extracts the page number and size of a request. Default (1, 0) if not set
+func getPageNumberAndSize(req *http.Request) (int, int) {
+	pageNumber := req.FormValue("page")
+	pageSize := req.FormValue("size")
+	numberInt, err := strconv.Atoi(pageNumber)
+	if err != nil || numberInt < 1 {
+		numberInt = 1
+	}
+	sizeInt, _ := strconv.Atoi(pageSize)
+	// Safeguard for negative numbers
+	if sizeInt < 0 {
+		sizeInt = 0
+	}
+	return numberInt, sizeInt
+}
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+func getChartList(repo string, pageNumber, pageSize int) (apiListResponse, interface{}, error) {
 	db, closer := dbSession.DB()
 	defer closer()
 	var charts []*models.Chart
-	if err := db.C(chartCollection).Find(nil).Sort("name").All(&charts); err != nil {
+
+	var query datastore.Query
+	c := db.C(chartCollection)
+	if repo != "" {
+		query = c.Find(bson.M{"repo.name": repo})
+	} else {
+		query = c.Find(nil)
+	}
+	query = query.Sort("name")
+
+	// TODO(andresmgot): We should not query the charts just yet in case we paginate the response
+	// but we need to do so since it's not possible to filter & count charts (countDocuments) until
+	// https://github.com/kubeapps/common/issues/8 is addressed
+	err := query.All(&charts)
+	if err != nil {
+		return apiListResponse{}, 0, err
+	}
+
+	totalPages := 1
+	if pageSize != 0 {
+		// If a pageSize is given, returns only the the specified number of charts and
+		// the number of pages
+		length := len(charts)
+		totalPages = int(math.Ceil(float64(length) / float64(pageSize)))
+
+		// If the page number is out of range, return the last one
+		if pageNumber > totalPages {
+			pageNumber = totalPages
+		}
+
+		first := pageSize * (pageNumber - 1)
+		last := min(first+pageSize, length)
+		charts = charts[first:last]
+	}
+
+	return newChartListResponse(charts), meta{totalPages}, nil
```

Comment:
How do we plan to handle the de-duplication of charts from different repos? `newChartListResponse` will de-dupe by digest, but if not all charts are loaded, we won't be able to fully de-dupe. This might be okay because we will at least de-dupe the charts that are shown on the page, but charts between pages will be shown, but we should look into what this looks like and add a comment considering it.

Reply:
That's a good point, I missed it. The issue is that, if there is a duplicated chart we will be returning `size-1` charts which will be weird. I think it's better to filter unique charts before this point so we can properly calculate the amount of pages shown and return the proper number of charts.

Reply:
This makes using `count` as above difficult. We should spend some time looking into ways to craft a MongoDB query to filter unique charts.

Reply:
Okay, I was able to craft a query using the aggregation API that is able to return charts with distinct digests:

```
db.charts.aggregate(
  [
    {
      $addFields:
        {
          firstChartVersion: { $arrayElemAt: [ "$chartversions", 0 ] },
        }
    },
    {
      $group: {
        _id: "$firstChartVersion.digest",
        chart: { $first: "$$ROOT" }
      }
    },
    {
      $replaceRoot: {
        newRoot: "$chart"
      }
    }
  ]
)
```

Explanation:
1. addFields - This copies the first element of the chartversions array to a new field, firstChartVersion. This is necessary for the next pipeline, as I couldn't find a way of grouping by a value in the first object in an array.
2. group - This groups all charts by the digest in the latest chartversion, and so removes any duplicates. Because this will return a new document with `_id` set to the digest, this also adds a new field `chart` that contains the original object so we can recover it.
3. replaceRoot - This final stage just recovers the original object by resetting the root to the `chart` field we added in the group operation.

Now we can add a stage before the first stage for the query using the `$match` pipeline, and we can add a final stage of `$count` or `$skip` and `$limit` to do the pagination. We may also be able to retrieve the count and paginated elements in the same query to optimise, probably using another `$addFields` stage, but I haven't checked for that.

Is it worth doing this over loading everything in memory, figuring out distinct charts and paginating client (server)-side? As you said, we're generally going to be dealing with a small enough collection that it probably won't make a huge difference. Also, there is some added work in supporting https://godoc.org/github.com/globalsign/mgo#Pipe in our datastore interface.

Personally, I think it is worth it to delegate the hard work to MongoDB (which is probably faster at doing this). It is less code, and with comments, I would argue simpler to understand what's going on. The filtering and uniquifying happens at the same time rather than separately, and so the result can clearly be understood by just looking at the single query.

That said, I think we need to find out exactly where the bottleneck is to understand if pagination will actually speed up the Hub. If the bottleneck is server -> database, then using this aggregated query will really help, because we avoid fetching more charts than we need to from the database. If loading all charts is actually pretty fast, and the bottleneck lies in sending this information back to the browser, then implementing pagination server-side should be fine (or it doesn't really matter whether this happens server-side or database). If sending all chart information to the browser also happens to be fast, and Angular/the browser is just slow at rendering all the cards and loading the icons, then we're better off just paginating browser-side in Angular and not touching the server-side at all. @andresmgot would you be able to spend some time investigating where the bottleneck is so we can decide on what the best solution is?

Reply:
An update here:

 - Querying all the resources like we were doing is around 80 ms (in my local instance, minikube)
 - Querying just a page of 32 resources is around 20 ms
 - Counting all the documents (`db.charts.find(...).count()`) is only 2 ms so it's not relevant
 - Doing an aggregation of functions takes as you suggest takes 40 ms (still better than the first option)

I have checked that heaviest operations in the hub are (apart than the angular app loading) are the download of the app itself (~200 ms) and then the logo(s) fetching that can take up to 250 ms. In any case if we reduce a good chunk of the time taken in the chartsvc it will be noticeable. Also note that the data send in the request gets reduced from 170kb to 16kb.

I am still adapting the `datastore` library to support the `Pipe` method.

Reply:
Thanks for the research @andresmgot, this is very helpful! So, it sounds like we should go with the aggregation query to get an accurate count of unique charts whilst also paginating.

As I suspected, the major bottleneck is Angular loading all the charts, so reducing to 40ms sounds reasonable here as we're get a more noticeable increase when loading less charts in the frontend.